[Concurrency] Update async alternative attribute

etcwilde opened the pull request:
Updating the async alternative attribute to take the completion handler index and do some limited type-checking. 

I'm still working on this, but getting it up so that we can bike-shed the name. I'm still working on getting the DeclName of the async function replaced with a reference to the resolved decl so that it doesn't need to be redone by everything that follows. With that decl, we can do a bit more type-checking and verify that the parameters match up.
Not entirely sure how to do that if the completion handler returns a non-void type, but at least we can verify that completion handler takes the same type as the async alternative function returns.

Resolves part of: rdar://72868440

etcwilde commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -5513,3 +5514,51 @@ void TypeChecker::checkClosureAttributes(ClosureExpr *closure) {
+             "must be a function type");
+    return;
+  }
+
```
Should verify that the return type of the completion handler is a `Void` type.

bnbarham replied:
There's some other checks that might be worth doing as well - ie. the handler shouldn't be attributed with `@autoclosure`. We could probably also check that it has `@escaping`?

etcwilde replied:
Done

bnbarham commented on include/swift/AST/Attr.h:
```diff
@@ -2145,19 +2145,29 @@ class TransposeAttr final
 
-  HasAsyncAlternativeAttr(SourceLoc AtLoc, SourceRange Range)
-    : DeclAttribute(DAK_HasAsyncAlternative, AtLoc, Range, false) {}
+  /// DeclName of the async function in the attribute
+  const DeclNameRef AsyncFunctionName;
```
Is the plan to find the function in a separate PR?

bnbarham replied:
Ah, you said you're still working on it 👍. Happy for it to be in this PR/another.

etcwilde replied:
Yeah, I'm still working on grabbing the function. I'm going to wrap it into the PR so I don't have to keep rev'ing the module version number. If this feature is blocking though, I can get this cleaned up and do it later though.

bnbarham replied:
Nope, not blocking. Leaving in this PR is okay with me.

bnbarham commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -5513,3 +5514,51 @@ void TypeChecker::checkClosureAttributes(ClosureExpr *closure) {
+  // We have two function decls to worry about here for now.
+  // There will be a third for the found async function decl.
+  // We have the decl that the attribute is attached to
+  //  callingFunctionDecl
```
How do you feel about `attachedFunctionDecl` instead of `callingFunctionDecl`?

bnbarham commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -5513,3 +5514,51 @@ void TypeChecker::checkClosureAttributes(ClosureExpr *closure) {
+  }
+
+  const ParameterList *params = callingFunctionDecl->getParameters();
+  if (!params) // TODO: Figure out how to exercise this and emit an error
+    llvm_unreachable("The parameter list is a nullptr");
```
From what I can tell, it's always set. All callers that I saw assume that it is as well.

Would be nice to clean up the AST construction at some point (ie. it should be obvious that this is true, but it really isn't since setParameters is public). In general there seems to be a fair bit of assumed postconditions that should be enforced better.

bnbarham commented on include/swift/AST/Attr.h:
```diff
@@ -2145,19 +2145,29 @@ class TransposeAttr final
+  const SourceLoc CompletionHandlerIndexLoc;
+
+  /// The index of the completion handler
+  const uint64_t CompletionHandlerIndex;
```
ParameterList uses `size_t`, may as well use it here as well (even if they are the same)

bnbarham reviewed (commented):
Thanks for adding that Evan! In terms of names, I really have no strong preference between `hasAsyncAlternative`, Doug's suggested `completionHandlerAsync`, or something else. I'm probably leaning towards `completionHandlerAsync` over `hasAsyncAlternative` since that doesn't use any new language.

etcwilde commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -1570,42 +1570,68 @@ void Parser::parseAllAvailabilityMacroArguments() {
 
-  auto Value = P.getStringLiteralIfNotInterpolated(
-      Loc, ("argument of '" + AttrName + "'").str());
+  SourceLoc nameLoc = P.Tok.getLoc();
+  Optional<StringRef> asyncFunctionName = P.getStringLiteralIfNotInterpolated(
+      nameLoc, ("argument of '" + AttrName + "'").str());
```
I think there was discussion at some point in the last year of switching over to lower-casing the first letter of variable names in LLVM. We're also pretty inconsistent internally. The ParseDecl file mixes styles, though newer code appears to use lower first character. I can change it, but I'm getting mixed signals from the surrounding code. 🤷

bnbarham replied:
Yeah, I was really hoping that variable name change would go through. Not sure where it's at right now... I don't really mind either way.

etcwilde commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -1570,42 +1570,68 @@ void Parser::parseAllAvailabilityMacroArguments() {
     return nullptr;
+  }
+
+  if (!P.consumeIf(tok::comma)) {
```
I personally like being explicit for readability, but Swift seems to like implicit, so sure, I guess. Can do that.

etcwilde replied:
Done

etcwilde committed:
Replace @hasAsyncAlt with @completionHandlerAsync

This patch replaces the @hasAsyncAlternative attribute with
@completionHandlerAsync. The @completionHandlerAsync attribute takes the
function decl name of the async function and optionally the index of the
completion hander parameter in the function that it's attached to.
If the completion handler index is not provided, it's assumed to be the
last parameter in the parameter list.

We resolve the async function while typechecking the attribute. Before
resolving, we verify that the function the attribute is attached to
isn't async, that it has enough parameters to at least have the
indicated completion handler referenced by the index, that the
completion handler is an escaping non-auto-closure function that returns
Void.

The async function declaration resolution isn't perfect yet, but I want
to get this patch up and we can refine it later. It pulls all of the
delcs with the specified declname in the same context as the
function that the attribute is attached to. Going through that list, it
keeps any that are async functions. If there are none, we emit an error
saying that there are no viable functions, if there are multiple we emit
an error saying that the decl name is ambiguous, and if there is one
function, we keep that as the resolve async function declaration.

This does not take into account the data types of the completion handler
or the async function. There are some complexities to making this
mapping. Here are the pieces:
 - If the completion handler takes a single data type, the async
   function should return that type. (easy case)
 - If the completion handler takes a `Result<T, Error>`, the async
   function is a throwing function that returns a T. (Medium difficulty)
 - If the completion handler looks like `(T?, Error?) -> Void`, we have
   an ambiguous situation between the following async functions:
   - func foo() async throws -> T
   - func foo() async throws -> T?
   That is, we cannot tell whether the `T?` in the completion handler is
   optional because it will be nil on an error, or if it is intended to
   be optional.

This can be done later if it becomes a problem.

etcwilde committed:
Updating tests for @completionHandlerAsync

This patch updates the tests for the @completionHandlerAsync attribute.

The main test is broken into three chunks; parsing, typechecking, and
the decl assginment phases.

For parsing, we verify both happy cases, with an explicitly provided
completion handler index and one where it is omitted. We also verify the
various ways that parsing can fail, with ill-formed names, missing
parens etc....

The typechecking phase verifies that the function the attribute is
attached to is not async, that it takes a completion handler, that the
completion handler is a function type, is a function that returns Void,
is @escaping, and is not @autoclosure.

The decl assignment complains about the name not referencing an existing
function, and the ambiguous naming of async functions.

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
Huh, Linux doesn't pick up that `SmallVector` has a default size, but MacOS seems to. MacOS just has a failing test that I'm not sure is related. Not seeing it locally anyway.

bnbarham commented on test/attr/attr_completionhandlerasync.swift:
```diff
@@ -0,0 +1,129 @@
+// expected-error@+1{{'@completionHandlerAsync' attribute cannot be applied to this declaration}}
+@completionHandlerAsync("foo", completionHandlerIndex: 0)
+struct SomeStruct: SomeProto {
+  // expected-error@+1:27{{no suitable '@completionHandlerAsync' async functions 'protoFunc'}}
```
The message sort of makes it sound like there should be another `protoFunc` that is `async` and attributed with `@completionHandlerAsync`. What do you think about `no corresponding async function named 'protoFunc' for 'protoFunc'`? Maybe "async version of function" or "async alternative function" even.

bnbarham commented on test/attr/attr_completionhandlerasync.swift:
```diff
@@ -0,0 +1,129 @@
+// Typechecking
+// ===================
+
+// expected-error@+2:23{{'@completionHandlerAsync' attached to async function}}
```
Thoughts on `'@completionHandlerAsync' should be attached to non-async completion-handler function`? It could be used for the `... must take completion handler` one as well.

bnbarham replied:
Could add a note on the async to say "function is declared async"

bnbarham commented on test/attr/attr_completionhandlerasync.swift:
```diff
@@ -0,0 +1,129 @@
+@completionHandlerAsync("betterFunc")
+func typecheckFunc2() {}
+
+// expected-error@+2:36{{completion handler must be a function type}}
```
Could use the message I suggested above + a note on the completionHandler to say it's not a function type

bnbarham commented on test/attr/attr_completionhandlerasync.swift:
```diff
@@ -0,0 +1,129 @@
+@completionHandlerAsync("foo")
+func typecheckFunc4(value: String, completionHandler: String) {}
+
+// expected-error@+2:33{{completion handler must return 'Void'}}
```
Could use the message suggested above + a note on the Int

bnbarham commented on test/attr/attr_completionhandlerasync.swift:
```diff
@@ -0,0 +1,129 @@
+@completionHandlerAsync("foo", completionHandlerIndex: 2)
+func typecheckFunc6(value: Int) { }
+
+// expected-error@+3:21{{'@completionHandlerAsync' completion handler must be '@escaping'}}
```
Same, but with a note on the type

etcwilde commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -4552,24 +4552,34 @@ llvm::Error DeclDeserializer::deserializeDeclCommon() {
+            asyncFunctionName, /*functionNameLoc*/ SourceLoc(),
+            explicitHandlerIndex, handlerIndex, /*handlerIndexLoc*/ SourceLoc(),
+            /*atLoc*/ SourceLoc(), /*range*/ SourceRange());
+        reinterpret_cast<CompletionHandlerAsyncAttr *>(Attr)
```
Yeah, since I'm creating the thing on the statement above, we _know_ that the `Attr` variable contains that type.
The `reinterpret_cast` says that I know what I'm doing and want to cast the type without any conversions and is completely free. The `cast` on the other hand actually does runtime checking and isn't free.

That said, I think a `static_cast` will do the trick and would definitely be better than a yolo-cast.

etcwilde commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -5669,3 +5669,153 @@ void TypeChecker::checkClosureAttributes(ClosureExpr *closure) {
+    // Get the list of candidates based on the name
+    // Grab all functions that are async
+    // TODO: Sanity check types -- we just use the DeclName for now
+    //  - Need a throwing decl if the completion handler takes a Result type
```
Yeah, that's fair. This function is starting to get a little long. If it's alright with you, I'll leave it here for now until we need to pull it out though.

bnbarham replied:
Yep, sorry - my comment wasn't worded very well. I think this is fine for now, but I'll likely pull the common parts of this and the refactoring out when I get back to it.

etcwilde replied:
Ah, yeah. Sounds good. 👍

etcwilde commented:
@swift-ci please smoke test

etcwilde commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -5669,3 +5669,153 @@ void TypeChecker::checkClosureAttributes(ClosureExpr *closure) {
+
+      for (AbstractFunctionDecl *candidate : candidates) {
+        diagnose(candidate->getLoc(), diag::decl_declared_here,
+                 candidate->getName());
```
Maybe I'm missing something, but I didn't see a way of passing the decl into the diagnose without changing `decl_declared_here` defined in `DiagnosticsSema.def`, which I don't want to do since others are using it. This isn't too bad.

bnbarham replied:
That's okay, was just pointing it out since I only recently found it 👍

etcwilde commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -4552,24 +4552,34 @@ llvm::Error DeclDeserializer::deserializeDeclCommon() {
+            readRecord(scratch, explicitHandlerIndex, handlerIndex,
+                       asyncFunctionDeclID, isCompound, rawPieces);
+
+        DeclNameRef asyncFunctionName;
```
Yeah, it can probably be reconstituted from the resolved function decl now. Will look into that.

etcwilde replied:
For the AST printer, that's what emits the results for `-dump-ast`? I see some attributes (`@escaping`, e.g) showing up in types, but I don't see attributes attached to declarations showing up (`@inline(never)` or `@dynamicMemberLookup` e.g) showing up. Not sure what the answer is there.

I will check on the emitted swift interface though. once I remember how to do that. :)

bnbarham commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3337,6 +3337,33 @@ ERROR(diff_params_clause_param_not_differentiable,none,
 
+// completionHanderAsync attribute
+ERROR(attr_completion_handler_async_handler_not_func,none,
+      "'%0' should be attached to non-async completion-handler function",
```
Tiny nitpick: "to a non-async"

etcwilde replied:
Fixed

bnbarham commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3337,6 +3337,33 @@ ERROR(diff_params_clause_param_not_differentiable,none,
+     "%0 is not a function type", (Type))
+
+NOTE(note_attr_completion_handler_async_handler_attr_req,none,
+     "completion handler must%select{not|}0 be '@%1'",
```
`{ not|}` (added space), as a side note there's no test for the not case.

etcwilde replied:
Fixed

etcwilde committed:
Remove AsyncFunctionName from serialization

We can get the DeclName of the async function from the resolved function
decl itself, so we don't actually need to serialize it. This patch pulls
the declNameRef from the serialization.

etcwilde commented:
@swift-ci please smoke test

etcwilde committed:
Remove ExplicitCompletionHandlerIndex flag

Assuming that we don't typecheck a deserialized module, we don't
actually need the ExplicitCompletionHandlerIndex. We used this flag to
determine whether we could trust the number stored in the handler index
or just take the last parameter of the function we're attached to.

Since we only typecheck it before serialization, we can safely check
that the completion handler location is valid before choosing whether we
should trust it or not.

etcwilde commented:
@swift-ci please smoke test

bnbarham commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -1630,7 +1630,7 @@ parseCompletionHandlerAsyncAttribute(Parser &P, StringRef AttrName,
   return new (P.Context) CompletionHandlerAsyncAttr(
       parsedAsyncName.formDeclNameRef(P.Context), nameLoc,
-      handlerIndexLoc.isValid(), handlerIndex, handlerIndexLoc, AtLoc,
+      handlerIndex, handlerIndexLoc, AtLoc,
```
Just noticed that `handlerIndex` is uninitialized in the default case, probably worth setting it to something even if it'll end up being overridden in type checking. 0/max?

etcwilde replied:
Okay, zero initialized now.

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci please smoke test macOS platform

etcwilde commented:
@swift-ci please smoke test macOS platform

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci please smoke test Windows

compnerd commented:
@swift-ci please test Windows platform

compnerd commented:
@swift-ci please test Windows platform

etcwilde commented:
@swift-ci please smoke test Linux platform

etcwilde commented:
@swift-ci please smoke test macOS platform

bnbarham reviewed (approved):
LGTM. Thanks Evan!

etcwilde commented:
@swift-ci please smoke test Linux platform

etcwilde commented:
I'm so sad that this had a merge conflict as soon as everything passed. :(

@swift-ci please smoke test

etcwilde commented:
@swift-ci please smoke test macOS

etcwilde commented:
@swift-ci please smoke test macOS

etcwilde commented:
@swift-ci please smoke test Windows platform

etcwilde commented:
This passed on the Windows bot previously. It looks like the bot is just full. I feel comfortable merging.
